File: scripts/lib/scenario_manifest.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import re
from pathlib import Path, PurePosixPath
from typing import NoReturn


VALID_LANES = {"secretless", "provider-e2e"}
VALID_PLATFORMS = {"any", "linux", "macos"}
VALID_PROVIDERS = {"codex", "claude", "gemini"}
PERSONA_PATTERN = re.compile(r"^[a-z][a-z0-9-]*$")


def die(message: str) -> NoReturn:
    raise SystemExit(message)


def _require_nonempty_string(value: object, label: str) -> str:
    if not isinstance(value, str) or not value.strip():
        die(f"{label} must be a non-empty string")
    return value


def _optional_bool(scenario: dict[str, object], key: str, scenario_id: str, default: bool) -> bool:
    if key not in scenario:
        return default
    value = scenario[key]
    if not isinstance(value, bool):
        die(f"scenario {scenario_id}: {key} must be boolean")
    return value


def _normalize_test_file(value: object, scenario_id: str, manual: bool) -> str:
    if value in ("", None):
        if manual:
            return ""
        die(f"scenario {scenario_id}: test_file is required for automated scenarios")

    if not isinstance(value, str):
        die(f"scenario {scenario_id}: test_file must be a string")

    path = PurePosixPath(value)
    if path.is_absolute() or any(part in ("", ".", "..") for part in path.parts):
        die(f"scenario {scenario_id}: test_file must stay under tests/scenarios without traversal")
    if path.suffix != ".sh" or not path.name.startswith("test-"):
        die(f"scenario {scenario_id}: test_file must reference a scenario shell script")
    return path.as_posix()


def load_scenarios(manifest_path: Path) -> list[dict[str, object]]:
    try:
        manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        die(f"Scenario manifest does not exist: {exc.filename}")

    if not isinstance(manifest, dict):
        die("scenario manifest must contain a JSON object")
    if manifest.get("version") != 1:
        die("scenario manifest version must be 1")

    raw_scenarios = manifest.get("scenarios")
    if not isinstance(raw_scenarios, list) or not raw_scenarios:
        die("scenario manifest must contain a non-empty scenarios array")

    seen_ids: set[str] = set()
    seen_test_files: set[str] = set()
    scenarios: list[dict[str, object]] = []
    for index, raw in enumerate(raw_scenarios, start=1):
        if not isinstance(raw, dict):
            die(f"scenario entry {index} must be an object")

        scenario_id = _require_nonempty_string(raw.get("id"), f"scenario entry {index} id")
        if scenario_id in seen_ids:
            die(f"Duplicate scenario id: {scenario_id}")
        seen_ids.add(scenario_id)

        description = _require_nonempty_string(
            raw.get("description"), f"scenario {scenario_id} description"
        )
        persona = _require_nonempty_string(raw.get("persona"), f"scenario {scenario_id} persona")
        if not PERSONA_PATTERN.fullmatch(persona):
            die(f"scenario {scenario_id}: persona must match {PERSONA_PATTERN.pattern}")

        providers = raw.get("providers")
        if not isinstance(providers, list) or not providers:
            die(f"scenario {scenario_id}: providers must be a non-empty list")
        normalized_providers: list[str] = []
        for provider in providers:
            provider_name = _require_nonempty_string(provider, f"scenario {scenario_id} provider")
            if provider_name not in VALID_PROVIDERS:
                die(
                    f"scenario {scenario_id}: provider must be one of: {', '.join(sorted(VALID_PROVIDERS))}"
                )
            if provider_name in normalized_providers:
                die(f"scenario {scenario_id}: duplicate provider {provider_name}")
            normalized_providers.append(provider_name)

        requires_credentials = _optional_bool(raw, "requires_credentials", scenario_id, False)
        manual = _optional_bool(raw, "manual", scenario_id, False)

        lane = raw.get("lane", "provider-e2e" if requires_credentials else "secretless")
        if not isinstance(lane, str) or lane not in VALID_LANES:
            die(f"scenario {scenario_id}: lane must be one of: {', '.join(sorted(VALID_LANES))}")
        if lane == "secretless" and requires_credentials:
            die(f"scenario {scenario_id}: cannot be secretless and require credentials")

        platform = raw.get("platform", "any")
        if not isinstance(platform, str) or platform not in VALID_PLATFORMS:
            die(
                f"scenario {scenario_id}: platform must be one of: {', '.join(sorted(VALID_PLATFORMS))}"
            )

        test_file = _normalize_test_file(raw.get("test_file", ""), scenario_id, manual)
        if test_file:
            if test_file in seen_test_files:
                die(f"duplicate scenario test_file: {test_file}")
            seen_test_files.add(test_file)

        scenarios.append(
            {
                "id": scenario_id,
                "description": description,
                "persona": persona,
                "providers": normalized_providers,
                "requires_credentials": requires_credentials,
                "manual": manual,
                "lane": lane,
                "platform": platform,
                "test_file": test_file,
            }
        )

    return scenarios


def scenario_shell_tests(scenario_root: Path) -> set[str]:
    scenario_root = scenario_root.resolve()
    return {
        path.relative_to(scenario_root).as_posix()
        for path in sorted(scenario_root.rglob("test-*.sh"))
        if path.is_file()
    }


def verify_coverage(scenario_root: Path, manifest_path: Path) -> None:
    scenario_root = scenario_root.resolve()
    scenarios = load_scenarios(manifest_path)

    manifest_test_files = {str(scenario["test_file"]) for scenario in scenarios if scenario["test_file"]}
    for test_file in sorted(manifest_test_files):
        full_path = scenario_root / test_file
        if not full_path.is_file():
            die(f"Missing test file: tests/scenarios/{test_file}")

    orphaned_test_files = sorted(scenario_shell_tests(scenario_root) - manifest_test_files)
    if orphaned_test_files:
        joined = ", ".join(orphaned_test_files)
        die(f"Scenario scripts missing from manifest: {joined}")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Validate and enumerate Workcell scenario metadata.")
    subparsers = parser.add_subparsers(dest="command", required=True)

    list_parser = subparsers.add_parser("list-tsv", help="Print validated scenario rows as TSV.")
    list_parser.add_argument("manifest", type=Path)

    verify_parser = subparsers.add_parser(
        "verify-coverage", help="Validate the manifest and ensure shell test coverage parity."
    )
    verify_parser.add_argument("scenario_root", type=Path)
    verify_parser.add_argument("manifest", type=Path)
    return parser


def main(argv: list[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)

    if args.command == "list-tsv":
        for scenario in load_scenarios(args.manifest):
            requires_credentials = "1" if scenario["requires_credentials"] else "0"
            manual = "1" if scenario["manual"] else "0"
            print(
                "\t".join(
                    [
                        str(scenario["id"]),
                        str(scenario["test_file"]),
                        requires_credentials,
                        str(scenario["lane"]),
                        str(scenario["platform"]),
                        manual,
                    ]
                )
            )
        return 0

    if args.command == "verify-coverage":
        verify_coverage(args.scenario_root, args.manifest)
        return 0

    parser.error(f"unsupported command: {args.command}")
    return 2  # unreachable; defensive in case parser.error is overridden


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/scenarios/scenario_manifest.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Omkhar Arasaratnam

package scenarios

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strings"
)

var (
	validLanes     = map[string]struct{}{"secretless": {}, "provider-e2e": {}}
	validPlatforms = map[string]struct{}{"any": {}, "linux": {}, "macos": {}}
	validProviders = map[string]struct{}{"codex": {}, "claude": {}, "gemini": {}}
	personaPrefix  = "^[a-z][a-z0-9-]*$"
)

type Scenario struct {
	ID                  string
	Description         string
	Persona             string
	Providers           []string
	RequiresCredentials bool
	Manual              bool
	Lane                string
	Platform            string
	TestFile            string
}

func nonEmptyString(value any, label string) (string, error) {
	str, ok := value.(string)
	if !ok || strings.TrimSpace(str) == "" {
		return "", fmt.Errorf("%s must be a non-empty string", label)
	}
	return str, nil
}

func optionalBool(scenario map[string]any, key, scenarioID string, defaultValue bool) (bool, error) {
	value, ok := scenario[key]
	if !ok {
		return defaultValue, nil
	}
	boolValue, ok := value.(bool)
	if !ok {
		return false, fmt.Errorf("scenario %s: %s must be boolean", scenarioID, key)
	}
	return boolValue, nil
}

func normalizeTestFile(value any, scenarioID string, manual bool) (string, error) {
	if value == "" || value == nil {
		if manual {
			return "", nil
		}
		return "", fmt.Errorf("scenario %s: test_file is required for automated scenarios", scenarioID)
	}

	raw, ok := value.(string)
	if !ok {
		return "", fmt.Errorf("scenario %s: test_file must be a string", scenarioID)
	}

	if strings.HasPrefix(raw, "/") {
		return "", fmt.Errorf("scenario %s: test_file must stay under tests/scenarios without traversal", scenarioID)
	}

	parts := strings.Split(raw, "/")
	for _, part := range parts {
		if part == "" || part == "." || part == ".." {
			return "", fmt.Errorf("scenario %s: test_file must stay under tests/scenarios without traversal", scenarioID)
		}
	}
	if path.Ext(raw) != ".sh" || !strings.HasPrefix(path.Base(raw), "test-") {
		return "", fmt.Errorf("scenario %s: test_file must reference a scenario shell script", scenarioID)
	}
	return path.Clean(raw), nil
}

func loadManifest(pathname string) (map[string]any, error) {
	content, err := os.ReadFile(pathname)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, fmt.Errorf("Scenario manifest does not exist: %s", pathname)
		}
		return nil, err
	}

	var manifest any
	if err := json.Unmarshal(content, &manifest); err != nil {
		return nil, err
	}
	root, ok := manifest.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("scenario manifest must contain a JSON object")
	}
	return root, nil
}

func LoadScenarios(manifestPath string) ([]Scenario, error) {
	manifest, err := loadManifest(manifestPath)
	if err != nil {
		return nil, err
	}
	if manifest["version"] != float64(1) {
		return nil, fmt.Errorf("scenario manifest version must be 1")
	}

	rawScenarios, ok := manifest["scenarios"].([]any)
	if !ok || len(rawScenarios) == 0 {
		return nil, fmt.Errorf("scenario manifest must contain a non-empty scenarios array")
	}

	seenIDs := map[string]struct{}{}
	seenTestFiles := map[string]struct{}{}
	scenarios := make([]Scenario, 0, len(rawScenarios))
	for index, raw := range rawScenarios {
		entry, ok := raw.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("scenario entry %d must be an object", index+1)
		}

		scenarioID, err := nonEmptyString(entry["id"], fmt.Sprintf("scenario entry %d id", index+1))
		if err != nil {
			return nil, err
		}
		if _, ok := seenIDs[scenarioID]; ok {
			return nil, fmt.Errorf("Duplicate scenario id: %s", scenarioID)
		}
		seenIDs[scenarioID] = struct{}{}

		description, err := nonEmptyString(entry["description"], fmt.Sprintf("scenario %s description", scenarioID))
		if err != nil {
			return nil, err
		}
		persona, err := nonEmptyString(entry["persona"], fmt.Sprintf("scenario %s persona", scenarioID))
		if err != nil {
			return nil, err
		}
		if !isPersonaValid(persona) {
			return nil, fmt.Errorf("scenario %s: persona must match %s", scenarioID, personaPrefix)
		}

		rawProviders, ok := entry["providers"].([]any)
		if !ok || len(rawProviders) == 0 {
			return nil, fmt.Errorf("scenario %s: providers must be a non-empty list", scenarioID)
		}
		normalizedProviders := make([]string, 0, len(rawProviders))
		for _, provider := range rawProviders {
			providerName, err := nonEmptyString(provider, fmt.Sprintf("scenario %s provider", scenarioID))
			if err != nil {
				return nil, err
			}
			if _, ok := validProviders[providerName]; !ok {
				return nil, fmt.Errorf(
					"scenario %s: provider must be one of: %s",
					scenarioID,
					strings.Join(sortedKeys(validProviders), ", "),
				)
			}
			for _, existing := range normalizedProviders {
				if existing == providerName {
					return nil, fmt.Errorf("scenario %s: duplicate provider %s", scenarioID, providerName)
				}
			}
			normalizedProviders = append(normalizedProviders, providerName)
		}

		requiresCredentials, err := optionalBool(entry, "requires_credentials", scenarioID, false)
		if err != nil {
			return nil, err
		}
		manual, err := optionalBool(entry, "manual", scenarioID, false)
		if err != nil {
			return nil, err
		}

		lane := "secretless"
		if requiresCredentials {
			lane = "provider-e2e"
		}
		if rawLane, ok := entry["lane"]; ok {
			laneValue, ok := rawLane.(string)
			if !ok || !containsKey(validLanes, laneValue) {
				return nil, fmt.Errorf("scenario %s: lane must be one of: %s", scenarioID, strings.Join(sortedKeys(validLanes), ", "))
			}
			lane = laneValue
		}
		if lane == "secretless" && requiresCredentials {
			return nil, fmt.Errorf("scenario %s: cannot be secretless and require credentials", scenarioID)
		}

		platform := "any"
		if rawPlatform, ok := entry["platform"]; ok {
			platformValue, ok := rawPlatform.(string)
			if !ok || !containsKey(validPlatforms, platformValue) {
				return nil, fmt.Errorf(
					"scenario %s: platform must be one of: %s",
					scenarioID,
					strings.Join(sortedKeys(validPlatforms), ", "),
				)
			}
			platform = platformValue
		}

		testFile, err := normalizeTestFile(entry["test_file"], scenarioID, manual)
		if err != nil {
			return nil, err
		}
		if testFile != "" {
			if _, ok := seenTestFiles[testFile]; ok {
				return nil, fmt.Errorf("duplicate scenario test_file: %s", testFile)
			}
			seenTestFiles[testFile] = struct{}{}
		}

		scenarios = append(scenarios, Scenario{
			ID:                  scenarioID,
			Description:         description,
			Persona:             persona,
			Providers:           normalizedProviders,
			RequiresCredentials: requiresCredentials,
			Manual:              manual,
			Lane:                lane,
			Platform:            platform,
			TestFile:            testFile,
		})
	}

	return scenarios, nil
}

func containsKey(values map[string]struct{}, key string) bool {
	_, ok := values[key]
	return ok
}

func sortedKeys(values map[string]struct{}) []string {
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

func isPersonaValid(persona string) bool {
	if persona == "" || persona[0] < 'a' || persona[0] > 'z' {
		return false
	}
	for _, r := range persona[1:] {
		if r >= 'a' && r <= 'z' {
			continue
		}
		if r >= '0' && r <= '9' {
			continue
		}
		if r == '-' {
			continue
		}
		return false
	}
	return true
}

func ScenarioShellTests(scenarioRoot string) ([]string, error) {
	absRoot, err := filepath.Abs(scenarioRoot)
	if err != nil {
		return nil, err
	}
	if info, err := os.Stat(absRoot); err != nil || !info.IsDir() {
		return []string{}, nil
	}

	paths := []string{}
	if err := filepath.WalkDir(absRoot, func(current string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if entry.IsDir() {
			return nil
		}
		name := entry.Name()
		if !strings.HasPrefix(name, "test-") || filepath.Ext(name) != ".sh" {
			return nil
		}
		rel, err := filepath.Rel(absRoot, current)
		if err != nil {
			return err
		}
		paths = append(paths, filepath.ToSlash(rel))
		return nil
	}); err != nil {
		return nil, err
	}

	sort.Strings(paths)
	return paths, nil
}

func VerifyCoverage(scenarioRoot, manifestPath string) error {
	absRoot, err := filepath.Abs(scenarioRoot)
	if err != nil {
		return err
	}
	root, err := os.OpenRoot(absRoot)
	if err != nil {
		return err
	}
	defer root.Close()

	scenarios, err := LoadScenarios(manifestPath)
	if err != nil {
		return err
	}

	manifestTestFiles := map[string]struct{}{}
	for _, scenario := range scenarios {
		if scenario.TestFile != "" {
			manifestTestFiles[scenario.TestFile] = struct{}{}
		}
	}
	keys := make([]string, 0, len(manifestTestFiles))
	for key := range manifestTestFiles {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, testFile := range keys {
		info, err := root.Stat(filepath.FromSlash(testFile))
		if err != nil || !info.Mode().IsRegular() {
			return fmt.Errorf("Missing test file: tests/scenarios/%s", testFile)
		}
	}

	shellTests, err := ScenarioShellTests(absRoot)
	if err != nil {
		return err
	}
	orphaned := make([]string, 0)
	for _, testFile := range shellTests {
		if _, ok := manifestTestFiles[testFile]; !ok {
			orphaned = append(orphaned, testFile)
		}
	}
	if len(orphaned) > 0 {
		return fmt.Errorf("Scenario scripts missing from manifest: %s", strings.Join(orphaned, ", "))
	}
	return nil
}

func ListTSV(manifestPath string, w io.Writer) error {
	scenarios, err := LoadScenarios(manifestPath)
	if err != nil {
		return err
	}
	for _, scenario := range scenarios {
		requires := "0"
		if scenario.RequiresCredentials {
			requires = "1"
		}
		manual := "0"
		if scenario.Manual {
			manual = "1"
		}
		if _, err := fmt.Fprintln(w, strings.Join([]string{
			scenario.ID,
			scenario.TestFile,
			requires,
			scenario.Lane,
			scenario.Platform,
			manual,
		}, "\t")); err != nil {
			return err
		}
	}
	return nil
}

func Usage(program string) string {
	if program == "" {
		program = "workcell-scenario-manifest"
	}
	return fmt.Sprintf(
		"Usage: %s [list-tsv|verify-coverage] ...",
		program,
	)
}

func Run(program string, args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintln(stderr, Usage(program))
		return 2
	}

	switch args[0] {
	case "list-tsv":
		if len(args) != 2 {
			fmt.Fprintln(stderr, Usage(program))
			return 2
		}
		if err := ListTSV(args[1], stdout); err != nil {
			fmt.Fprintln(stderr, err)
			return 1
		}
		return 0
	case "verify-coverage":
		if len(args) != 3 {
			fmt.Fprintln(stderr, Usage(program))
			return 2
		}
		if err := VerifyCoverage(args[1], args[2]); err != nil {
			fmt.Fprintln(stderr, err)
			return 1
		}
		return 0
	default:
		fmt.Fprintln(stderr, Usage(program))
		fmt.Fprintf(stderr, "%s: unsupported command: %s\n", program, args[0])
		return 2
	}
}
